[jit] make jit/rpc share the same PythonFutureWrapper

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35039 [jit] make jit/rpc share the same PythonFutureWrapper**

This is the initial step towards merging ivalue future and rpc future

Differential Revision: [D20537164](https://our.internmc.facebook.com/intern/diff/D20537164)

wanchaol committed:
[jit] make jit/rpc share the same PythonFutureWrapper

This is the initial step towards merging ivalue future and rpc future

[ghstack-poisoned]

xush6528 commented on torch/testing/_internal/distributed/rpc/jit/rpc_test.py:
```diff
@@ -750,3 +750,20 @@ def test_user_rrefs_confirmed_remote(self):
+    @dist_init
+    def test_python_future_with_jit(self):
+        dst_rank = (self.rank + 1) % self.world_size
+        rref = self._create_rref()
```
This rref is not used.

xush6528 commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -410,35 +413,6 @@ If the future completes with an error, an exception is thrown.
-            }
-          },
-          py::call_guard<py::gil_scoped_release>());
 
```
Extra empty line.

wanchaol committed:
Update on "[jit] make jit/rpc share the same PythonFutureWrapper"


This is the initial step towards merging ivalue future and rpc future

Differential Revision: [D20537164](https://our.internmc.facebook.com/intern/diff/D20537164)

[ghstack-poisoned]

zhaojuanmao reviewed (changes requested):
overall looks good, would you please also add toPyObject() conversion from ivalue::Future? and merge "m.def("wait", [](PythonFutureWrapper& fut) {" ?

rohan-varma commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -589,12 +597,14 @@ inline IValue toIValue(
     } break;
+    case TypeKind::FutureType: {
+      return obj.cast<PythonFutureWrapper>().fut;
+    } break;
```
nit: break is not needed since we are returning?

xush6528 commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -232,6 +232,9 @@ PyObject* rpc_init(PyObject* /* unused */) {
+  // TODO Once python object can be tagged as IValue and c10::ivalue::Future is
+  // implemented as generic Future<IValue>, we can consider all rpc call
+  // to return a future<IValue> later on.
   auto future = shared_ptr_class_<FutureMessage>(module, "Future")
```
I am thinking about what Yanli is asking. 

This can be removed right, since "_invoke_rpc_torchscript" returns a `torch::jit::PythonFutureWrapper` now.

xush6528 commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -232,6 +232,9 @@ PyObject* rpc_init(PyObject* /* unused */) {
+  // TODO Once python object can be tagged as IValue and c10::ivalue::Future is
+  // implemented as generic Future<IValue>, we can consider all rpc call
+  // to return a future<IValue> later on.
   auto future = shared_ptr_class_<FutureMessage>(module, "Future")
```
I am thinking about Yanli's merging request.
This can only be merged after "_invoke_rpc_builtin" and "_invoke_rpc_python_udf" returns a `ivalue::Future` wrapped by `torch::jit::PythonFutureWrapper` directly rather than wrapping a `ivalue::Future` into a `utils::Future<Message>`?

wanchaol replied:
yeah will do that i a follow up PR

zhaojuanmao replied:
I mean to merge line 807 in torch/csrc/jit/python/init.cpp (somehow I can not inline comment on that line)

wanchaol committed:
Update on "[jit] make jit/rpc share the same PythonFutureWrapper"


This is the initial step towards merging ivalue future and rpc future

Differential Revision: [D20537164](https://our.internmc.facebook.com/intern/diff/D20537164)

[ghstack-poisoned]

wanchaol commented:
> overall looks good, would you please also add toPyObject() conversion from ivalue::Future? and merge "m.def("wait", [](PythonFutureWrapper& fut) {" ?

@zhaojuanmao for merging `wait`, I am doing a follow up PR to let PyRPCBuiltin return ivalue::Future

wanchaol committed:
Update on "[jit] make jit/rpc share the same PythonFutureWrapper"


This is the initial step towards merging ivalue future and rpc future

Differential Revision: [D20537164](https://our.internmc.facebook.com/intern/diff/D20537164)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] make jit/rpc share the same PythonFutureWrapper"


This is the initial step towards merging ivalue future and rpc future

Differential Revision: [D20537164](https://our.internmc.facebook.com/intern/diff/D20537164)

[ghstack-poisoned]

xush6528 commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -46,6 +46,26 @@
+      : fut(std::move(fut)) {}
+
+  IValue wait() {
+    fut->wait();
```
This line is new? Will this change trace behavior?

wanchaol replied:
no it will not, tested locally it works fine. Before this PR tracer directly return `fut->value()`, `wait` is a no-op in tracing mode.
